stan/math/opencl/gp_exp_quad_cov.hpp

```diff
@@ -0,0 +1,68 @@
+#ifndef STAN_MATH_OPENCL_GP_EXP_QUAD_COV_HPP
+#define STAN_MATH_OPENCL_GP_EXP_QUAD_COV_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/opencl/matrix_cl.hpp>
+#include <stan/math/opencl/kernels/gp_exp_quad_cov.hpp>
+#include <stan/math/opencl/err/check_matching_dims.hpp>
+#include <CL/cl.hpp>
+
+namespace stan {
+namespace math {
+/**
+ * Squared exponential kernel on the GPU.
+ *
+ * @param x input vector or matrix
+ * @param sigma standard deviation
+ * @param length_scale length scale
+ *
+ * @return Squared distance between elements of x.
+ */
+inline matrix_cl gp_exp_quad_cov(const matrix_cl& x, const double sigma,
+                                 const double length_scale) try {
```

Comment:
I personally like this more because it separates out the exception control flow from the code that actually does stuff.

> Also having separate declarations matrix_cl res(x.cols(), x.cols()); in try and catch blocks is redundant.

Because res is defined in the first try block the catch doesn't know about it

https://en.cppreference.com/w/cpp/language/function-try-block
> For all other functions, reaching the end of a catch clause is equivalent to return; if the function's return type is (possibly cv-qualified) void, otherwise the behavior is undefined.

I'm partially indifferent though so if you want it the other way that's fine by me as well

Reply:
@SteveBronder, I agree with @t4c1, if only for consistency. We really haven't used function try blocks and we shouldn't just introduce them here. Let's try to stay with the coding style of the rest of the Math library.

Action: remove function try and use narrowly scoped try/catch.

Reply:
I'll change this, but I also wanna push back on us not using function try blocks. Yes we don't normally use these but I really like the way they separate out the control flow. If we had multiple places where things needed to return something different given a particular part of the code failed then I get the narrowly defined try/catch block. But if the OpenCL code fails we always throw the same error and call to `check_opencl_error`. It feels like the function try-block makes things more clear in saying, "If things go wary in the OpenCL code we always call `check_opencl_error`."